src/transferslot.cpp

```diff
@@ -758,12 +769,19 @@ void TransferSlot::doio(MegaClient* client, DBTableTransactionCommitter& committ
                                     }
                                 }
 
-                                reqs[i]->prepare(finaltempurl.c_str(), transfer->transfercipher(),
-                                         &transfer->chunkmacs, transfer->ctriv,
-                                         asyncIO[i]->pos, npos);
+                                auto pos = asyncIO[i]->pos;
+                                auto req = reqs[i];    // shared_ptr so no object is deleted out from under the worker
+                                auto transferkey = transfer->transferkey;
+                                auto ctriv = transfer->ctriv;
+                                req->pos = pos;
+                                req->status = REQ_ENCRYPTING;
 
-                                reqs[i]->pos = ChunkedHash::chunkfloor(asyncIO[i]->pos);
-                                reqs[i]->status = REQ_PREPARED;
+                                client->mAsyncQueue.push([req, transferkey, ctriv, finaltempurl, pos, npos](SymmCipher& sc)
+                                    {
+                                        sc.setkey(transferkey.data());
+                                        req->prepare(finaltempurl.c_str(), &sc, ctriv, pos, npos); 
+                                        req->status = REQ_PREPARED;
+                                    });
```

Comment:
Can we guarantee that `req` is not accessed from another thread at the same time? Is it possible that `doio` runs again at the same time as `prepare` runs on the worker thread?

Reply:
I wonder if we need to serialize access to members of `req` to avoid race conditions (now and in the future).

Reply:
excellent question - doio() can run while prepare does, but the req status has been put to `req->status = REQ_ENCRYPTING` and doio() won't do anything else with that req until its status is put to `REQ_PREPARED` by the lambda.

Reply:
I was about to do a similar question about concurrent access to `HttpReqXfer` objects. Good to get the answer here already :smile: